Support AWS ARM instances in Lokomotive

margamanterola opened the pull request:
# Support AWS ARM instances in Lokomotive

This change adds an os_arch parameter for both controller and workers on AWS.

Commit:
assets/terraform-modules/aws: Add support for ARM64

This change adds a new os_arch parameter for both controller and workers
on AWS.

invidian commented on assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -141,15 +141,15 @@ storage:
         inline: |
           KUBELET_IMAGE_URL=quay.io/kinvolk/kubelet
-          KUBELET_IMAGE_TAG=v1.19.4
+          KUBELET_IMAGE_TAG=v1.19.4-${os_arch}
```
This is not needed. We publish arm64 compatible manfiests:
```console
$ docker manifest inspect quay.io/kinvolk/kubelet:v1.19.4
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1994,
         "digest": "sha256:91f6f7fca8e5c3bc466d855496563e13ad3bc115106227bd46145e3d4910110d",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1994,
         "digest": "sha256:3d74bf03a3d088ec04374148b087d895f309f3ae31b89e758239f54a2c157996",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}
```

marga-kinvolk replied:
I have no idea how this works :). I copied it from the packet module: https://github.com/kinvolk/lokomotive/blob/0b0c39c449f1b916182539f987de357e284b87a9/assets/terraform-modules/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl#L197

invidian replied:
Just remove the suffix and right image will be picked up automatically.

invidian commented on assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -141,15 +141,15 @@ storage:
       contents:
         inline: |
-          IMAGE_TAG=v3.4.14
+          IMAGE_TAG=v3.4.14${etcd_arch_tag_suffix}
```
etcd _should_ publish multiarch manifests from next release (https://github.com/etcd-io/etcd/pull/12630/commits/7f79ffb4092be1235d83a85657351145d3542405).

invidian commented on assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -193,7 +194,7 @@ storage:
             -v /etc/kubernetes:/etc/kubernetes:rw \
             --network=host \
-            quay.io/kinvolk/bootkube:v0.14.0-helm4 \
+            quay.io/kinvolk/bootkube:v0.14.0-helm4-${os_arch} \
```
our bootkube also offers multi-arch manifests.

marga-kinvolk replied:
Ok, I've removed this and checked that indeed it downloads the right image. I guess it should be removed from the packet code?

invidian replied:
Yes, but I guess it's out of scope, created an issue for it: #1379

invidian commented on assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -170,6 +170,7 @@ storage:
           ETCD_PEER_CERT_FILE=/etc/ssl/etcd/etcd/peer.crt
           ETCD_PEER_KEY_FILE=/etc/ssl/etcd/etcd/peer.key
           ETCD_PEER_CLIENT_CERT_AUTH=true
+          ${etcd_arch_options}
```
I think generic etcd_env_variables looping over string-string map would be better.

marga-kinvolk replied:
I've changed the name. I'm not sure what the string-string map would look like... Again, this is copied from the packet code and it probably shouldn't drift?

invidian replied:
Something like:
```
%{ for key, val in etcd_env_variables ~}
${key}=${val}
%{ endfor ~}
```

>Again, this is copied from the packet code and it probably shouldn't drift?

Hmm... Either way is fine then I guess.

invidian reviewed (changes requested):
We probably need a separate CI pipeline for AWS arm to be able to test components there.

invidian commented on pkg/platform/aws/aws.go:
```diff
@@ -35,6 +35,7 @@ type workerPool struct {
 	Count        int               `hcl:"count"`
 	SSHPubKeys   []string          `hcl:"ssh_pubkeys"`
 	InstanceType string            `hcl:"instance_type,optional"`
+	OSArch       string            `hcl:"os_arch,optional"`
```
This must also be documented.

marga-kinvolk replied:
I've added the parameters to docs/configuration-reference/platforms/aws.md, is there another place I should document this?

invidian replied:
I think only `docs/configuration-reference/platforms/aws.md` is fine, thanks!

Commit:
assets/terraform-modules/aws: add ignition hostname units

This change adds systemd units that set the hostname from the metadata
service to both the controllers and the workers in AWS.

These units won't be needed after the next Flatcar release, which will
include kinvolk/bootengine#21, which does a similar thing. This change
can be reverted after that.

invidian commented on assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:
```diff
@@ -12,6 +12,20 @@ systemd:
+        RemainAfterExit=yes
+        Type=oneshot
+        [Install]
+        RequiredBy=wait-for-dns.service
```
Why `wait-for-dns.service` requires it?

invidian commented on assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -37,6 +37,20 @@ systemd:
       enable: true
     - name: locksmithd.service
       mask: true
+    - name: sethostname.service
```
I think if we add this, we should have an e2e test for it to make sure hostname on matchine matches node name.

invidian commented on docs/configuration-reference/platforms/aws.md:
```diff
@@ -203,6 +203,7 @@ worker_pool "my-worker-pool" {
 | `asset_dir`                   | Location where Lokomotive stores cluster assets.                                                                                                                                                                                                                                                             | -               | string       | true     |
 | `cluster_name`                | Name of the cluster. **NOTE**: It must be unique per DNS Zone and region.                                                                                                                                                                                                                                    | -               | string       | true     |
 | `tags`                        | Optional details to tag on AWS resources.                                                                                                                                                                                                                                                                    | -               | map(string)  | false    |
+| `os_arch`                     | Flatcar Container Linux architecture to use (x86\_64, arm64).                                                                                                                                                                                                                                                | "x86\_64"        | string       | false    |
```
Actually, maybe we can use https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/ec2_instance_type#supported_architectures to detect the architecture automagically using Terraform, based on instance type?

surajssd commented:
@marga-kinvolk could you rebase and push again?
